File: csharp/sdk/samples/InterceptorServerSample/SampleInterceptors.cs
using System.Text.Json.Nodes;
using ModelContextProtocol.Interceptors.Protocol;
using ModelContextProtocol.Interceptors.Server;

/// <summary>
/// Sample interceptors demonstrating validation, mutation, and sink types.
/// </summary>
[McpServerInterceptorType]
public class SampleInterceptors
{
    /// <summary>
    /// Validates that tool call arguments don't contain PII patterns.
    /// </summary>
    [McpServerInterceptor(
        Name = "pii-validator",
        Description = "Checks tool call arguments for PII patterns",
        Type = InterceptorType.Validation,
        Events = [InterceptionEvents.ToolsCall],
        Phase = InterceptorPhase.Request)]
    public static ValidationInterceptorResult ValidatePii(JsonNode payload)
    {
        var json = payload.ToJsonString();

        // Simple pattern check for demonstration
        if (json.Contains("ssn", StringComparison.OrdinalIgnoreCase) ||
            json.Contains("social security", StringComparison.OrdinalIgnoreCase))
        {
            return ValidationInterceptorResult.Failure(
                new ValidationMessage
                {
                    Path = "$.arguments",
                    Message = "Payload may contain Social Security Number data",
                    Severity = ValidationSeverity.Error,
                });
        }

        return ValidationInterceptorResult.Success();
    }

    /// <summary>
    /// Redacts email addresses from tool call payloads.
    /// </summary>
    [McpServerInterceptor(
        Name = "email-redactor",
        Description = "Redacts email addresses from payloads",
        Type = InterceptorType.Mutation,
        Events = [InterceptionEvents.ToolsCall],
        Phase = InterceptorPhase.Request,
        PriorityHint = -1000)]
    public static MutationInterceptorResult RedactEmails(JsonNode payload)
    {
        var json = payload.ToJsonString();
        var redacted = System.Text.RegularExpressions.Regex.Replace(
            json,
            @"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}",
            "[EMAIL_REDACTED]");

        if (redacted != json)
        {
            return new MutationInterceptorResult
            {
                Modified = true,
                Payload = JsonNode.Parse(redacted),
            };
        }

        return new MutationInterceptorResult { Modified = false, Payload = payload };
    }

    /// <summary>
    /// Logs all intercepted events.
    /// </summary>
    [McpServerInterceptor(
        Name = "request-logger",
        Description = "Logs all requests",
        Type = InterceptorType.Sink,
        Events = [InterceptionEvents.All])]
    public static SinkInterceptorResult LogRequest(
        JsonNode payload,
        string @event,
        InterceptorPhase phase,
        InvokeInterceptorContext? context)
    {
        Console.Error.WriteLine($"[interceptor] event={@event} phase={phase} traceId={context?.TraceId ?? "none"} payloadSize={payload.ToJsonString().Length}");

        return new SinkInterceptorResult
        {
            Recorded = true,
            Metrics = new Dictionary<string, double>
            {
                ["payloadBytes"] = payload.ToJsonString().Length,
            },
        };
    }
}

File: typescript/sdk/examples/interceptor-server/src/sample-interceptors.ts
// Copyright 2025 The MCP Interceptors Authors. All rights reserved.

import {
  InterceptionEvents,
  validationFailure,
  validationSuccess,
  type RegisteredInterceptor,
} from '../../../dist/index.js';

function payloadText(payload: unknown): string {
  return JSON.stringify(payload);
}

export const sampleInterceptors: RegisteredInterceptor[] = [
  {
    descriptor: {
      name: 'pii-validator',
      description: 'Checks tool call arguments for PII patterns',
      type: 'validation',
      hooks: [{ events: [InterceptionEvents.ToolsCall], phase: 'request' }],
    },
    handler: (params) => {
      const json = payloadText(params.payload);
      if (/ssn|social security/i.test(json)) {
        return validationFailure(params.phase, {
          path: '$.arguments',
          message: 'Payload may contain Social Security Number data',
          severity: 'error',
        });
      }
      return validationSuccess(params.phase);
    },
  },
  {
    descriptor: {
      name: 'email-redactor',
      description: 'Redacts email addresses from payloads',
      type: 'mutation',
      hooks: [{ events: [InterceptionEvents.ToolsCall], phase: 'request' }],
      priorityHint: -1000,
    },
    handler: (params) => {
      const json = payloadText(params.payload);
      const redacted = json.replace(
        /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g,
        '[EMAIL_REDACTED]',
      );
      if (redacted === json) {
        return { type: 'mutation', phase: params.phase, modified: false, payload: params.payload };
      }
      return {
        type: 'mutation',
        phase: params.phase,
        modified: true,
        payload: JSON.parse(redacted) as unknown,
      };
    },
  },
  {
    descriptor: {
      name: 'request-logger',
      description: 'Logs intercepted events to stderr',
      type: 'sink',
      hooks: [
        { events: [InterceptionEvents.All], phase: 'request' },
        { events: [InterceptionEvents.All], phase: 'response' },
      ],
    },
    handler: (params) => {
      const size = payloadText(params.payload).length;
      const trace = params.context?.traceId ?? 'none';
      console.error(
        `[interceptor] event=${params.event} phase=${params.phase} traceId=${trace} payloadBytes=${size}`,
      );
      return {
        type: 'sink',
        phase: params.phase,
        recorded: true,
        metrics: { payloadBytes: size },
      };
    },
  },
];
